text/0088-metric-instrument-optional-refinements.md

```diff
@@ -0,0 +1,205 @@
+# Metric Instruments
+
+Removes the optional semantic declarations `Monotonic` and `Absolute`
+for metric instruments, declares the Measure and Observer instruments
+as _foundational_, and introduces a process for standardizing new
+instrument _refinements_.
+
+## Motivation
+
+With the removal of Gauge instruments and the addition of Observer
+instruments in the specification, the existing `Monotonic` and
+`Absolute` options began to create confusion.  For example, a Counter
+instrument is used for capturing changes in a Sum, and we could say
+that non-negative-valued metric events define a monotonic Counter, in
+the sense that its Sum is monotonic.  The confusion arises, in this
+case, because `Absolute` refers to the captured values, whereas
+`Monotonic` refers to the semantic output.
+
+From a different perspective, Counter instruments might be treated as
+refinements of the Measure instrument.  Whereas the Measure instrument
+is used for capturing all-purpose synchronous measurements, the
+Counter instrument is used specifically for synchronously capturing
+measurements of changes in a sum, therefore it uses `Add()` instead of
+`Record()`, and it specifies `Sum` as the standard aggregation.
+
+What this illustrates is that we have modeled this space poorly.  This
+does not propose to change any existing metrics APIs, only our
+understanding of the three instruments currently included in the
+specification: Measure, Observer, and Counter.
+
+## Explanation
+
+The Measure and Observer instrument are defined as _foundational_
+here, in the sense that any kind of metric instrument must reduce to
+one of these.  The foundational instruments are unrestricted, in the
+sense that metric events support any numerical value, positive or
+negative, zero or infinity.
+
+The distinction between the two foundational instruments is whether
+they are synchronous.  Measure instruments are called synchronously by
+the user, while Observer instruments are called asynchronously by the
+implementation.  Synchronous instruments (Measure and its refinements)
+have three calling patterns (_Bound_, _Unbound_, and _Batch_) to
+capture measurements.  Asynchronous instruments (Observer and its
+refinements) use callbacks to capture measurements.
+
+All measurement APIs produce metric events consisting of [timestamp,
+instrument descriptor, label set, and numerical
+value](api-metrics.md#metric-event-format).  Synchronous instrument
+events additionally have [Context](api-context.md), describing
+properties of the associated trace and distributed correlation values.
+
+Observer instruments have a well-defined _last value_ measured by the
+instrument that can be useful in defining aggregations.  To maintain
+this property, we impose this requirement: two or more calls to
+`Observe()` in a single Observer callback invocation are treated as
+duplicates of each other, and the last call to `Observe()` wins.
+
+Measure instruments do not define a _last value_ relationship.
```

Comment:
This is very helpful. Now that the last value aggregation is outlawed for measures, would you expect that the order that measurements are recorded within a collection interval shouldn't affect the aggregate result?

Reply:
My brain keeps tripping on this line in the OTEP. Would this imply that no one could write a last-value aggregation for a measure? Or just that we won't provide one out of the box, or something else?

Reply:
Discussed offline in the SIG. This basically means that since measures don't have a single point-in-time associated with them, last-value isn't a well-defined thing for measures. One could write a last-value aggregation, but it would have to just make some assumptions/assertions about what "last-value" means, which isn't something we want to define as a part of the SDK.

Reply:
[consider this resolved from my perspective...I can't resolve it myself]